### PR TITLE
KAFKA-10515: Properly initialize nullable Serdes with default values

### PR DESCRIPTION
**Pull request by thake**
This is a cherry pick of PR #9338 on branch 2.6.

Also introduced the notion of WrappingNullableSerdes (aligned to the concept
of WrappingNullableSerializer and WrappingNullableDeserializer) and centralized
initialization in WrappingNullables.

The added integeration test KTableKTableForeignKeyJoinDistributedTest tests
whether all serdes are now correctly set on all stream clients.

Reviewers: John Roesler <vvcephei@apache.org>

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    public static InternalProcessorContext asInternalProcessorContext(final ProcessorContext context) {
+        if (context instanceof InternalProcessorContext) {
+            return (InternalProcessorContext) context;
+        } else {
+            throw new IllegalArgumentException(
+                "This component requires internal features of Kafka Streams and must be disabled for unit tests."
+            );
+        }
+    }
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java]:
```suggestion
```

Let's not backport these methods, which are unused in 2.6. If we need them later, we can backport them at that time.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextUtils.java]: Note, there will also be an unused import after removing these methods. I don't know if you know, but you can run `./gradlew :streams:testAll` to fully run all the verifications locally.

**Review by vvcephei — COMMENTED** [review]
Thanks for the quick backport, @thake !

It looks good, but can we drop the unused backported methods ?

**Comment by vvcephei** [comment]
Ah, it also looks like a bunch of tests fail due to EasyMock expectations being needed, and I also saw a couple of the null checks failling.

**Comment by vvcephei** [comment]
Hey, @mimaison , do you think we can get this into 2.6.1?

**Comment by vvcephei** [comment]
ok to test

**Comment by vvcephei** [comment]
test this please

**Comment by vvcephei** [comment]
... I'm not sure how to get Jenkins to test this PR.

**Comment by mimaison** [comment]
@vvcephei Yes, feel free to merge in 2.6 once it's ready

**Comment by thake** [comment]
> Ah, it also looks like a bunch of tests fail due to EasyMock expectations being needed, and I also saw a couple of the null checks failling.

@vvcephei Actually I think that the tests are correct but the implementation incorrectly requires non null default serdes.

**Comment by thake** [comment]
Also fixed this error in #9478 for trunk. Should also be cherry picked to 2.7. Otherwise, every Stream topology without a default serde will fail if it uses foreign key joins. 

@vvcephei Can you review this and #9478? Thanks!

**Comment by vvcephei** [comment]
Ahh, I already commented on https://github.com/apache/kafka/pull/9478, but I think now I see the motivation for that PR.

If it was indeed just because there are internal tests that need new expectations now, we should just add the expectations instead of twisting the "main" code into handling an unexpected condition. By the way, I'm sure it's not obvious, but the fundamental reason for my taking this position is that in production code, the context serdes can never be null. There is a "default default" of ByteArraySerde even if you don't configure a "deafult" serde at all.

**Comment by thake** [comment]
@vvcephei Sorry for the confusion, creating a new commit now.

**Comment by thake** [comment]
Ready for review

**Review by vvcephei — APPROVED** [review]
Thanks for the update @thake !

In parallel, I've been working on https://github.com/apache/kafka/pull/9471 so that we can actually get PR builds on the 2.6 branch. I'm not sure exactly how it'll work; after I merge that PR, I might have to merge it in to your branch to get the tests to run. I just want to be a little extra cautious because @mimaison is currently releasing 2.6.1, so I'm trying not to break the build at the last minute.

Thanks again for all your help on this.

**Comment by thake** [comment]
@vvcephei feel free to rebase/merge this PR in any way that is needed. If I can be of any help, just let me know.

**Comment by vvcephei** [comment]
_Finally_ merged that Jenkinsfile PR (seeing a lot of timeouts, but decided just to go ahead and merge it).

Merged in the 2.6 branch, and now the tests are running! We should hopefully get to merge this today.

Thanks again, @thake

**Comment by vvcephei** [comment]
Merged to 2.6 (cc @mimaison )

Thanks again for the contribution, @thake !

**Pull request merged**